registry/remote/repository.go

```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+				httpMethod, dockerContentDigestHeader,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
 
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+		// GET without server `Docker-Content-Digest` header forces the
+		// expensive calculation
+		if calculatedDigest, err = calculateDigestFromResponse(resp, maxMetadataBytes); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		contentDigest = calculatedDigest
+	} else {
+		contentDigest = serverHeaderDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if refDigest != "" && refDigest != contentDigest {
+		return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
 	}
 
+	// 6. Finally, if we made it this far, then all is good; return.
 	return ocispec.Descriptor{
 		MediaType: mediaType,
 		Digest:    contentDigest,
-		Size:      size,
+		Size:      resp.ContentLength,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response, maxMetadataBytes int64) (digest.Digest, error) {
+	defer resp.Body.Close()
+
+	body := limitReader(resp.Body, maxMetadataBytes)
+	content, err := io.ReadAll(body)
+	if err != nil {
+		return "", fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+	resp.Body = io.NopCloser(bytes.NewReader(content))
+
+	return digest.FromBytes(content), nil
+}
+
 // verifyContentDigest verifies "Docker-Content-Digest" header if present.
 // OCI distribution-spec states the Docker-Content-Digest header is optional.
 // Reference: https://github.com/opencontainers/distribution-spec/blob/v1.0.1/spec.md#legacy-docker-support-http-headers
 func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
+	digestStr := resp.Header.Get(dockerContentDigestHeader)
+
+	if len(digestStr) == 0 {
 		return nil
 	}
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
-	}
-	if contentDigest != expected {
-		return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
+
+	if contentDigest, err := digest.Parse(digestStr); err == nil {
+		if contentDigest != expected {
+			return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
```

Comment:
nit: Maybe we can elaborate this error message a bit?

Reply:
Done.